lib/std/process/Child.zig

```diff
@@ -344,39 +344,36 @@ pub const RunResult = struct {
     stderr: []u8,
 };
 
-fn fifoToOwnedArrayList(fifo: *std.io.PollFifo) std.ArrayList(u8) {
+fn fifoToOwnedSlice(fifo: *std.io.PollFifo) ![]u8 {
     if (fifo.head != 0) fifo.realign();
-    const result = std.ArrayList(u8){
+    var result = std.ArrayListUnmanaged(u8){
         .items = fifo.buf[0..fifo.count],
         .capacity = fifo.buf.len,
-        .allocator = fifo.allocator,
     };
     fifo.* = std.io.PollFifo.init(fifo.allocator);
-    return result;
+    return result.toOwnedSlice(fifo.allocator);
 }
 
+pub const Output = struct {
+    stdout: []u8,
+    stderr: []u8,
+};
+
 /// Collect the output from the process's stdout and stderr. Will return once all output
 /// has been collected. This does not mean that the process has ended. `wait` should still
 /// be called to wait for and clean up the process.
 ///
 /// The process must be started with stdout_behavior and stderr_behavior == .Pipe
-pub fn collectOutput(
+fn collectOutput(
     child: ChildProcess,
-    stdout: *std.ArrayList(u8),
-    stderr: *std.ArrayList(u8),
+    /// Used for `stdout` and `stderr`.
+    allocator: Allocator,
```

Comment:
Since this is going to break some people's code, let's put a little effort into making it more future-proof. My suggestion also happens to be more familiar to existing callsites.

* Make them both `std.ArrayListUnmanaged`
* Make the function only append to them, not clear them

Sure, this implementation uses that fifo abstraction but that's not necessarily how it has to work. And append-only array list is generally a very flexible API from the caller's perspective.

I'll note also that this patch introduces more error handling that can be deleted with the above suggestion.

Reply:
Pushed a simple version of this API while I look into improving it (it currently still uses `std.io.poll` and then just calls `appendSlice` to copy the data from the FIFO to the ArrayListUnmanaged).

Would you like to see `std.io.poll` updated to use `ArrayListUnmanaged` as well (from a brief look I didn't see any particular reason a FIFO was chosen), or for `collectOutput` to use something more custom tailored to its use case, or did you have something else in mind?

Reply:
Pushed an attempt at making `std.io.Poller` not FIFO-specific and then using the `ArrayListUnmanaged`s directly here:

https://github.com/squeek502/zig/commit/b8c732741012a7483ba187dbffea2bf838b25005

Let me know if it looks like I'm vaguely on the right track and I'll add it to this PR.

(the CI failures are unrelated and are happening on master as well btw)

Reply:
Sorry I didn't look at your branch yet, I'll try to get to that later

Reply:
No worries, it's only a minor improvement after the `list.capacity == 0` optimization from https://github.com/ziglang/zig/pull/22691#discussion_r1940335218